### PR TITLE
mon/MonClient: resurrect original client_mount_timeout handling

### PR DESCRIPTION
**Pull request by idryomov**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by idryomov** [comment]
jenkins test api

**Comment by idryomov** [comment]
jenkins test windows

File: src/test/librados/misc.cc [hunk 1]
```diff
@@ -83,6 +83,26 @@ TEST(LibRadosMiscConnectFailure, ConnectFailure) {
...
+                              "AQAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAA=="));
+  ASSERT_EQ(0, rados_conf_set(cluster, "client_mount_timeout", "2s"));
+
+  utime_t start = ceph_clock_now();
+  ASSERT_EQ(-ETIMEDOUT, rados_connect(cluster));
+  utime_t end = ceph_clock_now();
+
+  utime_t dur = end - start;
+  ASSERT_GE(dur, utime_t(2, 0));
+  ASSERT_LT(dur, utime_t(4, 0));
```
Comment (rzarzynski) [inline src/test/librados/misc.cc:87]: ACK.

**Review by rzarzynski — APPROVED** [review]

**Comment by ljflores** [comment]
@idryomov I'm seeing a failure that looks related to this PR:

`RADOS timed out (error connecting to the cluster)`, `failed to fetch mon config`
/a/yuriw-2023-07-03_15:28:45-rados-wip-yuri-testing-2023-06-23-0831-distro-default-smithi/7324852
```
2023-07-03T21:07:34.736 INFO:tasks.workunit.client.0.smithi159.stderr:2023-07-03T21:07:34.731+0000 7fe36e1c9640  0 monclient(hunting): authenticate timed out after 300
2023-07-03T21:07:34.736 INFO:tasks.workunit.client.0.smithi159.stderr:[errno 110] RADOS timed out (error connecting to the cluster)
2023-07-03T21:07:34.740 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/mon-stretch/mon-stretch-fail-recovery.sh:53: TEST_stretched_cluster_failover_add_three_osds:  run_mgr td/mon-stretch-fail-recovery x
2023-07-03T21:07:34.740 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:553: run_mgr:  local dir=td/mon-stretch-fail-recovery
2023-07-03T21:07:34.740 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:554: run_mgr:  shift
2023-07-03T21:07:34.741 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:555: run_mgr:  local id=x
2023-07-03T21:07:34.741 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:556: run_mgr:  shift
2023-07-03T21:07:34.741 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:557: run_mgr:  local data=td/mon-stretch-fail-recovery/x
2023-07-03T21:07:34.741 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:559: run_mgr:  ceph config set mgr mgr_pool false --force
2023-07-03T21:12:34.866 INFO:tasks.workunit.client.0.smithi159.stderr:2023-07-03T21:12:34.862+0000 7fa4cb991640  0 monclient(hunting): authenticate timed out after 300
2023-07-03T21:12:34.867 INFO:tasks.workunit.client.0.smithi159.stderr:[errno 110] RADOS timed out (error connecting to the cluster)
2023-07-03T21:12:34.871 INFO:tasks.workunit.client.0.smithi159.stderr://home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:561: run_mgr:  get_asok_path
2023-07-03T21:12:34.871 INFO:tasks.workunit.client.0.smithi159.stderr://home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:116: get_asok_path:  local name=
2023-07-03T21:12:34.871 INFO:tasks.workunit.client.0.smithi159.stderr://home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:117: get_asok_path:  '[' -n '' ']'
2023-07-03T21:12:34.872 INFO:tasks.workunit.client.0.smithi159.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:120: get_asok_path:  get_asok_dir
2023-07-03T21:12:34.872 INFO:tasks.workunit.client.0.smithi159.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:108: get_asok_dir:  '[' -n '' ']'
2023-07-03T21:12:34.872 INFO:tasks.workunit.client.0.smithi159.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:111: get_asok_dir:  echo /tmp/ceph-asok.30235
2023-07-03T21:12:34.872 INFO:tasks.workunit.client.0.smithi159.stderr://home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:120: get_asok_path:  echo '/tmp/ceph-asok.30235/$cluster-$name.asok'
2023-07-03T21:12:34.873 INFO:tasks.workunit.client.0.smithi159.stderr://home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:561: run_mgr:  realpath /home/ubuntu/cephtest/clone.client.0/src/pybind/mgr
2023-07-03T21:12:34.886 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:561: run_mgr:  ceph-mgr --id x --osd-failsafe-full-ratio=.99 --debug-mgr 20 --debug-objecter 20 --debug-ms 20 --debug-paxos 20 --chdir= --mgr-data=td/mon-stretch-fail-recovery/x '--log-file=td/mon-stretch-fail-recovery/$name.log' '--admin-socket=/tmp/ceph-asok.30235/$cluster-$name.asok' --run-dir=td/mon-stretch-fail-recovery '--pid-file=td/mon-stretch-fail-recovery/$name.pid' --mgr-module-path=/home/ubuntu/cephtest/clone.client.0/src/pybind/mgr
2023-07-03T21:17:34.946 INFO:tasks.workunit.client.0.smithi159.stderr:failed to fetch mon config (--no-mon-config to skip)
2023-07-03T21:17:34.949 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:575: run_mgr:  return 1
2023-07-03T21:17:34.949 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/mon-stretch/mon-stretch-fail-recovery.sh:53: TEST_stretched_cluster_failover_add_three_osds:  return 1
```

The failure reproduced several other times in this test link:
http://pulpito.front.sepia.ceph.com/?branch=wip-yuri-testing-2023-06-23-0831

**Comment by ljflores** [comment]
Please re-add "needs-qa" when this needs testing again.

**Comment by idryomov** [comment]
@batrick and I agree that this appears to be caused by https://github.com/ceph/ceph/pull/50503 that was present in the same batch.

What I see happening is:
- `ceph mon set election_strategy connectivity` hangs for almost 3 hours and gets terminated:
```
2023-07-03T18:03:08.444 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/mon-stretch/mon-stretch-fail-recovery.sh:50: TEST_stretched_cluster_failover_add_three_osds:  ceph mon set election_strategy connectivity
2023-07-03T21:02:34.488 INFO:tasks.workunit.client.0.smithi159.stderr:Terminated
```
- teardown code is invoked which ostensibly kills daemons:
```
2023-07-03T21:02:34.521 INFO:tasks.workunit.client.0.smithi159.stderr://home/ubuntu/cephtest/clone.client.0/qa/standalone/mon-stretch/mon-stretch-fail-recovery.sh:1: TEST_stretched_cluster_failover_add_three_osds:  teardown td/mon-stretch-fail-recovery 1
2023-07-03T21:02:34.522 INFO:tasks.workunit.client.0.smithi159.stderr://home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:164: teardown:  local dir=td/mon-stretch-fail-recovery
2023-07-03T21:02:34.522 INFO:tasks.workunit.client.0.smithi159.stderr://home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:165: teardown:  local dumplogs=1
2023-07-03T21:02:34.522 INFO:tasks.workunit.client.0.smithi159.stderr://home/ubuntu/cephtest/clone.client.0/qa/standalone/ceph-helpers.sh:166: teardown:  kill_daemons td/mon-stretch-fail-recovery KILL
```
- however the test proceeds and attempts to run the next command which times out because there is nothing to connect to:
```
2023-07-03T21:02:34.627 INFO:tasks.workunit.client.0.smithi159.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/mon-stretch/mon-stretch-fail-recovery.sh:51: TEST_stretched_cluster_failover_add_three_osds:  ceph mon add disallowed_leader e
2023-07-03T21:07:34.736 INFO:tasks.workunit.client.0.smithi159.stderr:2023-07-03T21:07:34.731+0000 7fe36e1c9640  0 monclient(hunting): authenticate timed out after 300
2023-07-03T21:07:34.736 INFO:tasks.workunit.client.0.smithi159.stderr:[errno 110] RADOS timed out (error connecting to the cluster)
```
Please retest without https://github.com/ceph/ceph/pull/50503 in the mix.

**Comment by ljflores** [comment]
Sure, thanks for looking into it @idryomov

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcVbn2qwgo1793-wip-yuri-testing-2023-07-14-1641-old-wip-yuri-testing-2023-07-12-1332-old-wip-yuri-testing-2023-07-12-1140-old-wip-yuri-testing

**Pull request merged**